File: python/ineq_functions/cvalue.py
import numpy as np
from scipy.special import ndtri

from .helpers import get_bootstrap_indices


def base_sn(n: int, k: int, alpha: float) -> float:
    """Base function for the SN test statistic defined in eq (41) of
    Section 5 in Canay, Illanes, and Velez (2023). This function is called
    by the other cvalue functions. It is not exported. Function
    :func:`ineq_functions.cvalue.cvalue_sn` is a convenience wrapper that sets
    n and k based on the dimensions of the input matrix.

    Parameters
    ----------
    n : int
        Sample size.
    k : int
        Number of moments.
    alpha : float
        Significance level.

    Returns
    -------
    float
        The SN critical value.
    """
    # Obtain the quantile of the standard normal distribution corresponding to
    # the significance level
    z_quantile = ndtri(1 - alpha / k)

    # Compute the c-value as in eq (41)
    return z_quantile / np.sqrt(1 - z_quantile**2 / n)


def cvalue_sn(x_data: np.ndarray, alpha: float) -> float:
    """Calculate the c-value for the SN test statistic defined in eq (40) of
    Section 5 in Canay, Illanes, and Velez (2023). This is a convenience
    wrapper for :func:`ineq_functions.cvalue.base_sn` that sets n and k based
    on the dimensions of the input matrix.

    Parameters
    ----------
    x_data : array_like
        Matrix of the moment functions with n rows (output of
        :func:`ineq_functions.m_function`).
    alpha : float
        Significance level.

    Returns
    -------
    float
        The c-value for the SN test statistic.
    """
    n = x_data.shape[0]  # sample size
    k = x_data.shape[1]  # number of moments

    # Compute the c-value as in eq (41)
    return base_sn(n, k, alpha)


def cvalue_sn2s(x_data: np.ndarray, alpha: float, beta: float | None = None) -> float:
    """Calculate the c-value for the SN2S test statistic defined in eq (42) of
    Section 5 in Canay, Illanes, and Velez (2023).

    Parameters
    ----------
    x_data : array_like
        n x k matrix of the moment functions with n rows (output of
        :func:`ineq_functions.m_function`).
    alpha : float
        Significance level for the first stage test.
    beta : float, default: alpha / 50
        Significance level for the second stage test.

    Returns
    -------
    float
        The c-value for the SN2S test statistic.
    """
    n = x_data.shape[0]  # sample size
    k = x_data.shape[1]  # number of moments

    if beta is None:
        beta = alpha / 50

    # Step 1: define set J_sn as almost binding
    ## Run the first stage from cvalue_sn
    cvalue0 = base_sn(n, k, beta)

    ## Compute the mean of each column of x_data
    mu_hat = x_data.mean(axis=0)
    ## Compute the standard deviation of each column of x_data
    sigma_hat = x_data.std(axis=0)

    ## Studentized statistic for each moment inequality
    test_stat0 = np.sqrt(n) * mu_hat / sigma_hat

    ## Number of moment inequalities that are almost binding as in eq (40)
    k_hat = (test_stat0 > (-2 * cvalue0)).sum()

    # Step 2: calculate critical value using a subset of moment inequalities
    if k_hat > 0:
        return base_sn(n, k_hat, alpha - 2 * beta)  # as in eq (42)
    else:
        return 0


def cvalue_eb2s(
    x_data: np.ndarray,
    alpha: float,
    beta: float | None = None,
    bootstrap_replications: int | None = None,
    rng_seed: int | None = None,
    bootstrap_indices: np.ndarray | None = None,
) -> float:
    """Calculate the c-value for the EB2S test statistic defined in eq (49) of
    Section 5 in Canay, Illanes, and Velez (2023).

    Parameters
    ----------
    x_data : array_like
        n x k matrix of the moment functions with n rows (output of
        :func:`ineq_functions.m_function`).
    alpha : float
        Significance level for the first stage test.
    beta : float, default: alpha / 50
        Significance level for the second stage test.
    bootstrap_replications : int, optional
        Number of bootstrap replications. Required if bootstrap_indices
        is not specified.
    rng_seed : int, optional
        Random number generator seed (for replication purposes). If not
        specified, the system seed will be used as-is.
    bootstrap_indices : array_like, optional
        Integer array of shape (bootstrap_replications, n) for the bootstrap
        replications. If this is specified, bootstrap_replications and rng_seed
        will be ignored. If this is not specified, bootstrap_replications is
        required.

    Returns
    -------
    float
        The c-value for the EB2S test statistic.
    """
    n = x_data.shape[0]  # sample size

    if beta is None:
        beta = alpha / 50

    ## Step 1: Algorithm of the Empirical Bootstrap as in Section 5.2

    # Obtain random numbers for the bootstrap
    bootstrap_indices = get_bootstrap_indices(
        n, bootstrap_replications, rng_seed, bootstrap_indices
    )

    ## Compute the mean of each column of x_data
    mu_hat = x_data.mean(axis=0)
    ## Compute the standard deviation of each column of x_data
    sigma_hat = x_data.std(axis=0)

    x_samples = x_data[bootstrap_indices, :]

    ## Follow the steps in eq (46)
    web_matrix = (
        np.sqrt(n)
        * (1 / n)
        * (x_samples - mu_hat[np.newaxis, np.newaxis, :]).sum(axis=1)
        / sigma_hat[np.newaxis, :]
    )

    ## Take maximum of each sample
    web_vector = web_matrix.max(axis=1)

    ## Obtain quantile of bootstrap samples
    cvalue0 = np.quantile(web_vector, 1 - beta)

    ## Studentized statistic for each moment inequality
    test_stat0 = np.sqrt(n) * mu_hat / sigma_hat

    # Step 2: Critical value

    ## Selection of moment inequalities that are almost binding as in eq (47)
    almost_binding = test_stat0 > (-2 * cvalue0)
    if np.any(almost_binding):
        web_matrix2 = web_matrix[:, almost_binding]
        web_vector2 = web_matrix2.max(axis=1)
        # We use the midpoint interpolation method for consistency with MATLAB
        return np.quantile(web_vector2, 1 - alpha + 2 * beta, interpolation="midpoint")
    else:
        return 0

File: r/ineq_functions/R/cvalue.R
#' Stage 1 Critical Value
#'
#' @param x_data an n x k matrix of evaluated moment functions.
#' @param alpha the confidence level required.
#'
#' @return a float for the critical value.
#' @export
cvalue_sn <- function(x_data, alpha) {
  ## Step 1: parameter setting
  n <- nrow(x_data) # sample size
  k <- ncol(x_data) # number of moments

  ## Step 2: calculations
  qq <- qnorm(1 - alpha / k)
  # as in eq (40)
  c_sn <- qq / sqrt(1 - qq^2 / n)

  c_value <- c_sn
  return(c_value)
}

#' Stage 2 Critical Value
#'
#' @param x_data an n x k matrix of evaluated moment functions.
#' @param alpha the confidence level required for the first stage test.
#' @param beta the confidence level required for the second stage test.
#'
#' @return a float for the critical value.
#' @export
cvalue_sn2s <- function(x_data, alpha, beta) {
  ## Step 0: parameter setting
  n <- nrow(x_data) # sample size

  if (beta < alpha / 2) {
    #     cat('Two step SN-method is running');
  } else {
    stop(sprintf(
      "beta is not lower than alpha/2: fix it! alpha: %s, beta %s",
      alpha,
      beta
    ))
  }

  # Step 1: define set J_SN as almost binding
  ## Run the first stage from cvalue_sn
  cvalue0 <- cvalue_sn(x_data, beta) # as in eq (40)

  ## Compute the sample mean of each column of x_data
  mu_hat <- Rfast::colmeans(x_data)

  ## Compute the standard deviation of each column of x_data
  ## normalize by n instead of n-1 as in matlab code
  sigma_hat <- Rfast::colVars(x_data, std = T) * sqrt((n - 1) / n)

  ## Studentized statistic for each moment inequality
  test_stat0 <- sqrt(n) * mu_hat / sigma_hat

  ## Number of moment inequalities that are almost binding as in eq (39)
  k_hat <- sum(test_stat0 > (-2 * cvalue0))

  ## Step 2: calculate critical value using a subset of moment inequalities

  if (k_hat > 0) {
    qq1 <- qnorm(1 - (alpha - 2 * beta) / k_hat)
    # as in eq (41)
    cvalue1 <- qq1 / sqrt(1 - qq1^2 / n)
  } else {
    cvalue1 <- 0
  }

  return(cvalue1)
}

#' Bootstrap Critical Value
#'
#' @param x_data an n x k matrix of evaluated moment functions.
#' @param alpha the confidence level required for the first stage test.
#' @param beta the confidence level required for the second stage test.
#' @param bootstrap_replications the number of bootstrap replications.
#'   Required if bootstrap_indices is not specified.
#' @param rng_seed the seed for replication purposes. If not specified, the
#'   seed is not set.
#' @param bootstrap_indices an integer vector of indices to use for the
#'   bootstrap. If this is specified, bootstrap_replications and rng_seed will
#'   be ignored. If this is not specified, bootstrap_replications is required.
#'
#' @return a float for the critical value.
#' @export
cvalue_eb2s <- function(x_data, alpha, beta, bootstrap_replications = NULL, rng_seed = NULL, bootstrap_indices = NULL) {
  # sample size
  n <- nrow(x_data)
  # number of moments
  k <- ncol(x_data)

  if (beta >= alpha / 2) {
    stop(sprintf(
      "beta is not lower than alpha/2: fix it! alpha: %s, beta %s",
      alpha,
      beta
    ))
  }

  # Step 1: Algorithm of the Empirical Bootstrap as in Section 5.2

  ## Compute the mean of each column of x_data
  mu_hat <- Rfast::colsums(x_data) / n
  ## Compute the standard deviation of each column of x_data
  ## (normalize by n instead of n-1 as in MATLAB code)
  sigma_hat <- Rfast::colVars(x_data, std = T) * sqrt((n - 1) / n)

  # to replicate results
  if (is.null(bootstrap_indices)) {
    if (is.null(bootstrap_replications)) {
      stop(
        "bootstrap_replications must be specified if bootstrap_indices is not."
      )
    } else {
      if (!is.null(rng_seed)) {
        set.seed(rng_seed, kind = "Mersenne-Twister")
      }
      # Generate indices if not specified
      bootstrap_indices <- sample.int(
        n,
        n * bootstrap_replications,
        replace = T
      )
    }
  }

  # Indexing trick, the inner matrix has structure:
  # [ Col1 from BB1, Col2 from BB1, ... Col1 from BB2, Col2 from BB2, ... ]
  x_means <- matrix(
    Rfast::colsums(matrix(x_data[bootstrap_indices, ], nrow = n)) / n,
    ncol = k
  )
  # Follow the steps in eq (45)
  # (double transpose is because R broadcasts column vectors)
  WEB_matrix <- t(
    (sqrt(n) * (t(x_means) - mu_hat)) / sigma_hat
  )

  ## Take maximum of each sample
  WEB_vector <- Rfast::rowMaxs(WEB_matrix, value = T)

  ## Obtain quantile of bootstrap samples
  cvalue0 <- Rfast2::Quantile(WEB_vector, 1 - beta)

  ## Studentized statistic for each moment inequality
  test_stat0 <- sqrt(n) * mu_hat / sigma_hat

  # Step 2: Critical value

  ## Selection of moment inequalities that are almost binding as in eq (46)
  almost_binding <- (test_stat0 > (-2 * cvalue0))
  if (any(almost_binding)) {
    WEB_matrix2 <- WEB_matrix[, almost_binding]
    WEB_vector2 <- Rfast::rowMaxs(WEB_matrix2, value = T)
    cvalue1 <- Rfast2::Quantile(WEB_vector2, 1 - alpha + 2 * beta)
  } else {
    cvalue1 <- 0
  }

  return(cvalue1)
}
